librad2/src/project.rs

```diff
@@ -0,0 +1,60 @@
+use std::fmt;
+use std::fmt::Display;
+use std::ops::Deref;
+use std::str::FromStr;
+
+use crate::git;
+use crate::keys::device;
+use crate::meta::project as meta;
+use crate::paths::Paths;
+
+pub trait Project {
+    type ProjectId: FromStr + Display;
```

Comment:
Not sure what you mean by default implementation, but this trait is indeed not the right abstraction yet: when calling `init`, you know which type of project you want, but showing, listing, and even editing should be polymorphic over the backend. I guess `Project` needs to be an enum which just has a fixed (non-trait) impl

Reply:
Ah sorry, should have been more clear. What I mean is that why would you constrain the implementation of this trait with this associated type having these traits? I would only expect that if a default implementation was given using those traits.

re: abstraction. I think I had something similar with a repo and its history. I tied getting a repository along with history manipulation initially, but separated the two. Maybe you'd want something similar. So something like `trait ProjectInit`, and `trait ProjectCommands`

Reply:
Because at some point I thought `impl Trait` would permit runtime subtyping, but then I wrote the top level `ProjectId` enum because that’s actually not true. Nevermind

Reply:
Alrighty :+1: